File: src/main/java/kitchenpos/products/tobe/application/DefaultProductService.java
package kitchenpos.products.tobe.application;

import kitchenpos.products.tobe.infra.ProductEntity;
import kitchenpos.products.tobe.dto.ProductDto;
import kitchenpos.products.tobe.exception.ProductDuplicationException;
import kitchenpos.products.tobe.infra.ProductRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class DefaultProductService implements ProductService {

    private final ProductRepository repository;

    public DefaultProductService(final ProductRepository repository){
        this.repository = repository;
    }

    @Transactional
    @Override
    public ProductDto register(ProductDto dto){
        validateRegisteredProduct(dto.getName());

        ProductEntity productEntity = new ProductEntity.Builder()
            .id(dto.getId())
            .name(dto.getName())
            .price(dto.getPrice())
            .build();

        ProductEntity savedProductEntity = repository.save(productEntity);

        return new ProductDto(savedProductEntity);
    }

    @Transactional(readOnly = true)
    @Override
    public List<ProductDto> list(){
        return repository.list()
            .stream()
            .map(product -> new ProductDto(product))
            .collect(Collectors.toList());
    }

    protected void validateRegisteredProduct (final String name){
        if(repository.findByName(name)){
            throw new ProductDuplicationException("동일한 상품이 이미 등록 되었습니다.");
        }
    }

}

File: src/main/java/kitchenpos/products/tobe/application/DefaultProductService.kt
package kitchenpos.products.tobe.application

import kitchenpos.products.application.ProductService
import kitchenpos.products.domain.Product
import kitchenpos.products.domain.ProductRepository
import kitchenpos.products.tobe.domain.CreateProductService
import kitchenpos.products.tobe.dto.PriceChangeEvent
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Transactional
@Service("ProductService")
class DefaultProductService(
    private val productRepository: ProductRepository,
    private val createProductService: CreateProductService,
    private val changeProductPriceService: ChangeProductPriceService,
    private val priceChangeEventPublisher: PriceChangeEventPublisher,
) : ProductService {
    override fun create(request: Product): Product {
        val product = createProductService.createProduct(
            price = request.price,
            name = request.name,
        )

        return productRepository.save(product)
    }

    override fun changePrice(
        productId: UUID,
        request: Product,
    ): Product {
        val product = changeProductPriceService.changeProductPrice(
            productId = productId,
            price = request.price,
        )

        priceChangeEventPublisher.publishPriceChange(PriceChangeEvent(productId))

        return product
    }

    @Transactional(readOnly = true)
    override fun findAll(): List<Product> = productRepository.findAll()
}
